mgr/dashboard: Improve prettier scripts and documentation

tspmelo opened the pull request:
Added prettier validation to the run-frontend-unittests script. This will make
sure we are always running prettier in our commits.

Added 2 new npm scripts:
- 'prettier', will run prettier formatter on all frontend files
- 'prettier:lint', will check all frontend files against prettier linter

Removed 'pretty-quick' and related scripts. Since we now have all files
prettified we can simply run prettier on them.

Remove 'tslint-eslint-rules' package and all related rules. Prettier can check
all the removed rules.

Updated HACKING.rst with some information about prettier.

Prettify all remaining files.

Add reference to prettier in CodingStyle.

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/app.component.ts:
```diff
@@ -8,7 +8,7 @@ import { AuthStorageService } from './shared/services/auth-storage.service';
   selector: 'cd-root',
   templateUrl: './app.component.html',
-  styleUrls: ['./app.component.scss']
+  styleUrls: ['./app.component.scss'],
```
For me, it's ok to go with trailing commas
but Prettier:
npm run makePretty
erases them... What should we do about it?

tspmelo replied:
This comma was added just to trigger the prettier validation and test it in jenkins. It will be removed before the PR is merged.

I personally think its cleaner to not use trailing commas and we can use the same style for TS and JSON arrays/objects. 

Currently prettier is configured to remove all trailing commas. We can discuss a change to this option in the future, but changing it is not in the context of this PR.

tspmelo replied:
@alfonsomthd For now I would suggest to remove it.

tspmelo commented on src/pybind/mgr/dashboard/frontend/package.json:
```diff
@@ -10,7 +10,7 @@
     "lint": "ng lint",
     "e2e": "ng e2e",
-    "makePretty": "pretty-quick --staged"
+    "makePretty": "pretty-quick"
```
Not sure if you are talking about the script name or the plugin we use, so:
- We use `makePretty` so we know we are not using prettify directly. 
  But maybe calling it `pretty-quick` would make more sense.
  In the past this was a pre-commit git hook.
- We are using `pretty-quick` because it allow us to format just the files modified on your current branch, ignoring all the old files that are still not well formated and are irelevant to the PR. This plugin uses prettier.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/styles/popover.scss:
```diff
@@ -16,7 +16,7 @@
 }
 
-.message {
+.messssage {
```
The commit where this change is was pushed only for testing reasons, it will be removed before merging.

tspmelo commented on src/pybind/mgr/dashboard/run-frontend-unittests.sh:
```diff
@@ -21,4 +21,7 @@ if [ -e ${config}_old ]; then
   mv ${config}_old $config
 fi
 
+a=() ; for l in $(git diff --name-status master | grep  [.ts\|.scss]  | sed -n 's/.*frontend\///p'); do a+=($l); done ; files="{$(echo ${a[*]} | tr ' ' ',')}"
```
It seems I really need to change the cmd, since it didn't work in jenkins.
I will consider your cmd when I make this change, thanks.

tspmelo commented:
jenkins retest this please

tspmelo commented:
It seems to be working now.
If the prettier linter finds any problem it will output the following:
![image](https://user-images.githubusercontent.com/399326/43071578-2cb75858-8e6b-11e8-8872-4e0089674c74.png)
It will list all files with problems and then some helper text.

tspmelo commented:
jenkins retest this please

tspmelo commented:
I have changed this PR.

The script I added in the unittest script was not working well and could cause problems depending in your git configuration. The same applies to running `pretty-quick`.

Given those problems, I decided to finally fix all linting problems and just use the prettier CLI to check and format the code.

tspmelo commented:
This should only be merged after #22695 , to reduce conflicts.

votdev commented on CodingStyle:
```diff
@@ -155,8 +155,10 @@ For Angular code, we follow the official Angular style guide:
 
-    https://www.npmjs.com/package/codelyzer
+    https://palantir.github.io/tslint/
+    http://codelyzer.com/
```
https://codelyzer.com/ works, but Firefox does not like the SSL certificate. Chrome 68 will warn you if you browse to a not SSL encrypted site.

tspmelo replied:
Codelyzer website doesn't have a valid SSL certificate.
Even they, on github [1], send you to the http site.

I think it's better to have chrome just showing a non intrusive warning, over all browsers preventing you from accessing the site before you accept the risk and add an exception to the browser configuration.

[1] https://github.com/mgechev/codelyzer

tspmelo commented:
jenkins retest this please
